pandas/core/groupby.py

```diff
@@ -3740,9 +3740,31 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        # GH 13204
+        # Here, categorical groupers are contained:
+        # - in `result` as columns (unexpanded)
+        # - in `index` as level values (already expanded).
+        # Multi-index `self.grouper.result_index` is an unexpaned version of
+        # `index` and could be a valid index for the current `result`.
+        # `result.index` is (and should be) flat.
+        #
+        # Expand `result` and clean its index
+        indexer = self.grouper.result_index.get_indexer(index)
+        result = result.reindex(index=indexer, copy=False)
+        result.reset_index(drop=True, inplace=True)
+        # Grouper columns got garbled with NaN's,
```

Comment:
that was a note before I looked at doing the set/reset. More concerned with idiomatic / correct / parsimoniuos code here. The comment can be added, but I still want to have a plain english expln for future readers why this is necessary.